homeassistant/components/device_tracker/icloud.py

```diff
@@ -19,49 +20,116 @@
 CONF_INTERVAL = 'interval'
 DEFAULT_INTERVAL = 8
 
+# entity attributes
+ATTR_USERNAME = 'username'
+ATTR_PASSWORD = 'password'
+ATTR_ACCOUNTNAME = 'accountname'
 
+ICLOUDTRACKERS = {}
+
+
+# def setup_scanner(hass, config, see, add_devices_callback):
 def setup_scanner(hass, config, see):
     """ Set up the iCloud Scanner. """
-    from pyicloud import PyiCloudService
-    from pyicloud.exceptions import PyiCloudFailedLoginException
-    from pyicloud.exceptions import PyiCloudNoDevicesException
 
     # Get the username and password from the configuration
     username = config.get(CONF_USERNAME)
     password = config.get(CONF_PASSWORD)
+    name = config.get(CONF_NAME, username)
+
+    iclouddevice = Icloud(hass, username, password, name)
+    ICLOUDTRACKERS['icloud.' + name] = iclouddevice
+    # add_devices_callback([iclouddevice])
+
+    def lost_iphone(call):
+        """ Calls the lost iphone function if the device is found """
+        accountname = call.data.get('accountname')
+        entity_id = 'icloud.' + accountname
+        devicename = call.data.get('devicename')
+        if entity_id in ICLOUDTRACKERS:
+            ICLOUDTRACKERS[entity_id].lost_iphone(devicename)
+
+    hass.services.register('device_tracker', 'lost_iphone',
+                           lost_iphone)
+
+    def update_icloud(call):
+        """ Calls the update function of an icloud account """
+        accountname = call.data.get('accountname')
+        entity_id = 'icloud.' + accountname
+        if entity_id in ICLOUDTRACKERS:
+            ICLOUDTRACKERS[entity_id].update_icloud(see)
+
+    hass.services.register('device_tracker',
+                           'update_icloud', update_icloud)
+    iclouddevice.update_icloud(see)
 
-    if username is None or password is None:
-        _LOGGER.error('Must specify a username and password')
-        return False
+    track_utc_time_change(
+        hass, iclouddevice.update_icloud(see),
+        minute=range(0, 60, config.get(CONF_INTERVAL, DEFAULT_INTERVAL)),
+        second=0
+    )
+
+    return True
 
-    try:
-        _LOGGER.info('Logging into iCloud Account')
-        # Attempt the login to iCloud
-        api = PyiCloudService(username,
-                              password,
-                              verify=True)
-    except PyiCloudFailedLoginException as error:
-        _LOGGER.exception('Error logging into iCloud Service: %s', error)
-        return False
 
-    def keep_alive(now):
-        """ Keeps authenticating iCloud connection. """
-        api.authenticate()
-        _LOGGER.info("Authenticate against iCloud")
+class Icloud(Entity):  # pylint: disable=too-many-instance-attributes
```

Comment:
There is no need to have this extend from an `Entity` as the device tracker does not work with entities. You can just extend from `object`.